### PR TITLE
Advanced Search functionality #632

### PR DESCRIPTION
**Pull request by tomaszandel**
https://github.com/cyfronet-fid/eosc-search-service/issues/632

### CODE REVIEW

**Comment by eldenkat** [comment]
Steps to reproduce the behavior:
1. Go to 'search service'
2. Enter "batteries" in the main bar
3. Go to "narrow your search" bar.
4. Pick "author" and add the name "aldrich" 
5. See one item.

6. Change "all catalogs" to "software" in the drop-down list.
7. See "no data" output (correct)                     
8. Leave software in drop-down list and pick software from the upper bar 
9. See 58 outputs                             

10. Change "software" to "all catalogs" in drop-down list.
11. See the saved query, and yet the result is 2632 items                                     


 

**Expected behavior**
Fix the bug

**Comment by eldenkat** [comment]
Steps to reproduce the behavior:
1. Go to 'search service'
2. Enter "syndrome" in the main bar
3. Go to "Show Advanced Search".
4. Enter "down" phrase and pick "exact match" from the drop-down list
5. See that two first items doesn't contain "down" phrase.                                                                                


6. Expected bahaviour: proper fix.

**Comment by Marcelinna** [comment]
the button doesn't clear
image

**Comment by Marcelinna** [comment]
Adding items works only when you press the "Enter" button, "Enter" clicked on the keyboard does not work
image

**Comment by Marcelinna** [comment]
Lack of text highlighting when I user advance serach
- Advance Search
image
- Main Search
image

**Comment by Marcelinna** [comment]
1. Go to serach 
2. Type to main serach "DATA"
3. Go to advance search
4. Type exact: single molecules
image
5. See problem - result 2 publication, only in one publication I see the word single molecules
image

**Comment by Marcelinna** [comment]
1. Go to Search
2. Go to Advance Search
3. Type any of: Bordeaux
image
4. Change to "none of" and click "Enter" button
5. See problem - the change is not implemented
image

**Comment by tomaszandel** [comment]
> 

In Down case, both have DOWN , we are searchign in all fields for Down, its case insensitive as fields in solr are indexed as case insensitive

**Comment by tomaszandel** [comment]
> 1. Go to Search
> 2. Go to Advance Search
> 3. Type any of: Bordeaux
>    image
> 4. Change to "none of" and click "Enter" button
> 5. See problem - the change is not implemented
>    image

This one ise invalid, you have to add it to narrow, in that case as you can see you have only any of in filter list

**Comment by tomaszandel** [comment]
> Lack of text highlighting when I user advance serach
> 
> * Advance Search
>   image
> * Main Search
>   image

There will be no highlighting for filters, as solr does not support it

**Comment by tomaszandel** [comment]
> 1. Go to Serach
> 2. Type to main search "DATA"
> 3. Go to advanced search
> 4. Type exact: single molecules
>    image
> 5. See problem - result 2 publication, only in one publication I see the word single molecules
>    image

You have it also in description, solr does not accept '-' characters so we are stripping them off, thus it matches

**Comment by tomaszandel** [comment]
> the button doesn't clear image

it clears the narrow :)) Button is always associated with inpuit

**Comment by tomaszandel** [comment]
> Steps to reproduce the behavior:
> 
> 1. Go to 'search service'
> 2. Enter "batteries" in the main bar
> 3. Go to "Show Advanced Search".
> 4. Pick "author" and add the name "aldrich"
> 5. See one item.
>    
> 6. Change "all catalogs" to "software" in the drop-down list.
> 7. See "no data" output (correct)                     
> 8. Leave software in drop-down list and pick software from the upper bar
> 9. See 58 outputs
>    
> 10. Change "software" to "all catalogs" in drop-down list.
> 11. See the saved query, and yet the result is 2632 items
>     
> 
> **Expected behavior** Fix the bug or improve user journey.

This si valid,  standard search has been executed, will fix it

**Comment by Marcelinna** [comment]
1. Type w the main search 
2. Type in advance search exact: down
image
3. I don't see word "down" in found publications
(e.g title [AN INTERESTING CASE OF RAISED RIGHT SIDED PRESSURE – WHAT IS IT?, [H63D Syndrome renamed Oslo Syndrome]
image

**Comment by Marcelinna** [comment]
@tomaszandel check go back to serach

**Comment by Marcelinna** [comment]
1. Add parameters to filter
image
2. Refresh the page
image
3. Click on the button "show advance search"
image
4. See on the url ---> 
https://search-6.docker-fid.grid.cyf-kr.edu.pl/search/software?q=*&standard=false&tags=author:%20test&tags=exact:%20chemistry

All parameters are stored in url, but the elements are not visible after showing the advance search

**Comment by Marcelinna** [comment]
image
only one publication
image
two publications
image

**Comment by eldenkat** [comment]
Related to the @Marcelinna comment above"
Task: Find publications authored by "Bernard Legras" on the topic of "climate change". Result:


There is one publication by Bernard Legras related to climate change, but it cannot be found:

**Comment by eldenkat** [comment]
Not sure it could be improved, but:

Steps to reproduce:
1. Go to advanced search
2. Enter "dead sea" and pick "exact"                                                       


3. See results - "dead sea" in title is only 10th 



If irrelevant, I'll move the comment. But doesn't seems user friendly to show results with matching keywords first.

**Comment by Marcelinna** [comment]
@tomaszandel When I type author and some space
image

**Comment by Marcelinna** [comment]
@tomaszandel Change Hide Advanced Search to Hide and Clear Advanced Search

**Comment by eldenkat** [comment]
General guidelines for AS:

1. Change author names to "author name"

2. Higlights - in each catalog.
3. To decide and correlate with other collections: consider whether the filter once applied (e.g. restricted acess) should disappear when changing the category (e.g. between data and software), whether it should be kept and show no data if it is not applicable in a given category.
4. Change of UI - move drop-down with collections higher (@aleksandrahojszyk )

**Comment by aleksandrahojszyk** [comment]
filters should be cleared between categories because they are not applicable between each

**Comment by Marcelinna** [comment]
1. Enter phase to the main bar
2. See problem - clear button show in Advance Search too
image

File: backend/app/routes/web/search.py [hunk 1]
```diff
@@ -80,6 +80,67 @@ async def search_post(
...
+    """
+    Do a search against the specified collection.
+
+    The q, qf, fq, sort params correspond to
+    https://solr.apache.org/guide/8_11/query-syntax-and-parsing.html.
+    Paging is cursor-based, see
+    https://solr.apache.org/guide/8_11/pagination-of-results.html#fetching-a-large-number-of-sorted-results-cursors.
+    """
+    final_solr_sorting = await define_sorting(sort_ui, sort)
+    print("advanced search triggered")
```
Comment (Michal-Kolomanski) [inline backend/app/routes/web/search.py:118]: Print statement to remove

**Review by michal-szostak — APPROVED** [review]
I have a couple of minors, but generally it's alright.

**Review by eldenkat — APPROVED** [review]

**Comment by michal-szostak** [comment]
And one suggestion - I would change naming from `searchadv` to `search_advanced`. `adv` reminds me of ads - which if we follow this naming scheme may at some point create issue with content blockers. I think a little bit more explicity would be better.

File: backend/app/routes/web/search.py (hunk 1 above)
Comment (tomaszandel) [reply backend/app/routes/web/search.py]: done

**Pull request merged**